### PR TITLE
Add tp_fastnew and tp_fastinit slots for faster object instanciation

### PR DESCRIPTION
**Pull request by haypo**
This branch is based on the tp_fastcall PR #65.

It's still a work-in-progress, not ready for a review. I wanted to created a pull request to start the discussion on the bug tracker, especially ask if the optimization is worth it.

Changes:

* Add tp_fastnew and tp_fastinit slots to PyTypeObject
* Implement tp_fastnew and tp_fastinit for list and _asyncio.Future types: just to run a few benchmarks
* Implement @fastinit decorator in Argument Clinic
* Add slot_tp_fastnew() and slot_tp_fastinit(): call __new__() / __init__() using FASTCALL calling convention
* Support FASTCALL in wrapper object (wrapper_call): add PyWrapperFlag_FASTCALL flag
* Add other wrappers/glue code in typeobject.c to support calling tp_new or tp_fastnew (and tp_init or tp_fastinit), support inheritance, support backward compatibility with the stable ABI, etc. => this part is still a WIP and may have some bugs

### CODE REVIEW

File: Lib/test/test_sys.py [hunk 1]
```diff
@@ -1092,7 +1092,7 @@ def delx(self): del self.__x
         check((1,2,3), vsize('') + 3*self.P)
         # type
         # static type: PyTypeObject
-        fmt = 'P2n15Pl4Pn9Pn11PIP'
+        fmt = 'P2n15Pl4Pn9Pn11PIPPPP'
```
Comment (methane) [inline Lib/test/test_sys.py]: s/PPPP/4P/

File: Objects/abstract.c [hunk 2]
```diff
@@ -2249,7 +2249,18 @@ PyObject_Call(PyObject *callable, PyObject *args, PyObject *kwargs)
...
 
-    result = (*call)(callable, args, kwargs);
+    if (PyType_HasFeature(Py_TYPE(callable), Py_TPFLAGS_HAVE_FASTCALL)
+       && Py_TYPE(callable)->tp_fastcall) {
+        result = _Py_RawFastCallDict(callable,
+                                     Py_TYPE(callable)->tp_fastcall,
+                                     &PyTuple_GET_ITEM(args, 0),
+                                     PyTuple_GET_SIZE(args),
+                                     kwargs);
+        result = _Py_CheckFunctionResult(callable, result, NULL);
```
Comment (methane) [inline Objects/abstract.c]: _Py_CheckFunctionResult() is called twice. (later in this function).

File: Objects/methodobject.c [hunk 3]
```diff
@@ -87,34 +87,38 @@ PyCFunction_Call(PyObject *func, PyObject *args, PyObject *kwargs)
...
-    /* _PyMethodDef_RawFastCallDict() must not be called with an exception set,
-       because it can clear it (directly or indirectly) and so the
+    /* _PyCFunction_FastCallDict() must not be called with an exception set,
+       because it may clear it (directly or indirectly) and so the
        caller loses its exception */
     assert(!PyErr_Occurred());
 
-    assert(method != NULL);
+    assert(PyCFunction_Check(func_obj));
+    assert(func_obj != NULL);
```
Comment (methane) [inline Objects/methodobject.c]: The second assert seems redundant.

File: Objects/rangeobject.c [hunk 4]
```diff
@@ -69,18 +69,16 @@ make_range_object(PyTypeObject *type, PyObject *start,
...
-    if (!_PyArg_NoKeywords("range()", kw))
+    if (!_PyArg_NoStackKeywords("range()", kwnames))
         return NULL;
 
-    if (PyTuple_Size(args) <= 1) {
-        if (!PyArg_UnpackTuple(args, "range", 1, 1, &stop))
-            return NULL;
-        stop = PyNumber_Index(stop);
+    if (nargs == 1) {
+        stop = PyNumber_Index(args[0]);
```
Comment (methane) [inline Objects/rangeobject.c]: Wouldn't this change changes error message when range() called without no argument?

File: Objects/rangeobject.c [hunk 5]
```diff
@@ -96,7 +94,7 @@ range_new(PyTypeObject *type, PyObject *args, PyObject *kw)
         }
     }
     else {
-        if (!PyArg_UnpackTuple(args, "range", 2, 3,
+        if (!_PyArg_UnpackStack(args, nargs, "range", 2, 3,
```
Comment (methane) [inline Objects/rangeobject.c]: Maybe, we can use `1, 3` instead of `2, 3`.

File: Lib/test/test_sys.py (hunk 1 above)
Comment (haypo) [inline Lib/test/test_sys.py]: Done

File: Objects/rangeobject.c (hunk 4 above)
Comment (haypo) [inline Objects/rangeobject.c]: Right, I fixed the error message. But it was already wrong before :-)

File: Objects/abstract.c (hunk 2 above)
Comment (haypo) [inline Objects/abstract.c]: Fixed

**Comment by haypo** [comment]
Ok, all issues that you reported Naoki should be fixed now ;-)

**Comment by haypo** [comment]
Thanks for the review.

**Comment by brettcannon** [comment]
Closing this to make sure I didn't miss notifying anyone about the migration.

**Pull request closed without merging**